### PR TITLE
fix: Initial setup for pouch adapter

### PR DESCRIPTION
**Pull request by ninetteadhikari**
## Current setup
Currently PouchDB uses a single adapter to store all application data in one place including the core data as well as the view index data. The adapter for dependent databases is registered in `registerDependentDatabase` in the `adapter.js` file.

## Issue
The use of a single adapter per database becomes a problem when the user wants to store their core data and view index data separately in multiple databases. The current setup does not provide the option to do so. 

## Use case
One of the use cases for using separate adapters can include using the `memory` adapter for indexes to speed up queries. Separate adapters is also helpful in taking advantage of the encryption options of the different databases.

## Approach
This PR provides an option for the user to specify a separate adapter to store their view index in addition to the main adapter for their core data.
For example the user will be able to define their main adapter using `adapter` and view index adapter using `view_adapter`. Here the core data will be stored in IndexedDB and view index data in WebSQL.
```
const db = new PouchDB(DB_NAME, {adapter: 'idb', view_adapter: 'websql'})
```

### CODE REVIEW

**Comment by janl** [comment]
> The adapter is registered…

The adapter for dependent databases is registered…

Other uses cases could include using the `memory` adapter for indexes to speed up queries.

**Comment by garbados** [comment]
Would it be better to establish a broader syntax for customizing dependent DBs? For example:

```javascript
const db = new PouchDB('example', {
  dependentDbs: {
    adapter: 'websql',
    auto_compaction: true,
    PouchDB: CustomPouchDB
  }
})
```

Right now we ONLY modify dependent DB options if the `view_adapter` option has been set. This difference does not have to have a large code footprint:

```javascript
let dependentDbConstructor = this.constructor
let dependentDbOptions = clone(this.__opts)
if (dependentDbOptions.dependentDbs) {
  const { dependentDbs, ...baseOpts, PouchDB: CustomPouchDB } = dependentDbOptions
  if (CustomPouchDB) { dependentDbConstructor = CustomPouchDB }
  dependentDbOptions = { ...baseOpts, ...dependentDbs }
}
var dependentDb = new dependentDbConstructor(dependentDb, dependentDbOptions);
```

What do you think?

File: tests/integration/test.viewadapter.js [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+    afterEach(function (done) {
+      testUtils.cleanup([dbs.name], done);
+    });
+
+    it('Create pouch with separate view adapters', function (done) {
+      var db = new PouchDB(dbs.name, {adapter: 'idb', view_adapter: 'memory'});
+
+      db.bulkDocs(docs).then(function () {
+        db.put(ddoc).then(function () {
+          db.query('index', {
```
Comment (garbados) [inline tests/integration/test.viewadapter.js:58]: This snippet exhibits the [pyramid of doom](https://en.wikipedia.org/wiki/Pyramid_of_doom_(programming)). We should be using async/await syntax instead, like this:

```javascript
async function () {
  await db.bulkDocs({ docs: [ ...docs, ddoc] })
  // ...
}
```

File: tests/integration/test.viewadapter.js [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+            viewRequest.onupgradeneeded = function (event) {
+              // The version of the view database created is 1 which shows that this
+              // database was newly created in IndexedDB and did not exist there
+              // before. So the view database was created in the database specified in
+              // the view_adapter and not in the default `idb`adapter.
+              event.oldVersion.should.equal(0);
+              event.newVersion.should.equal(1);
+            };
+
+            viewRequest.onsuccess = function () {
```
Comment (garbados) [inline tests/integration/test.viewadapter.js:77]: This callback is not guaranteed to have been run by the end of this test. We should await its completion before continuing, in order to guarantee our assertions actually occur.

File: tests/integration/test.viewadapter.js [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+        });
+      });
+    });
+
+    it('Create pouch with no view adapters', function (done) {
+      var db = new PouchDB(dbs.name, {adapter: 'idb'});
+
+      db.bulkDocs(docs).then(function () {
+        db.put(ddoc).then(function () {
+          db.query('index', {
```
Comment (garbados) [inline tests/integration/test.viewadapter.js:100]: Why perform this query if you don't do anything with the result?

File: tests/integration/test.viewadapter.js [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+
+    afterEach(function (done) {
+      testUtils.cleanup([dbs.name], done);
+    });
+
+    it('Create pouch with separate view adapters', function (done) {
+      var db = new PouchDB(dbs.name, {adapter: 'idb', view_adapter: 'memory'});
+
+      db.bulkDocs(docs).then(function () {
+        db.put(ddoc).then(function () {
```
Comment (garbados) [inline tests/integration/test.viewadapter.js:57]: Why put the docs and ddoc separately?

**Comment by garbados** [comment]
Just for some context, [other companies](https://github.com/RangerMauve/pouchdb-adapter-hyperbee/issues/1#issuecomment-843623485) have sought this feature and been dissuaded from using PouchDB by its absence. This is part of why I support a broader `dependentDbs` configuration syntax, rather than just a `view_adapter` option. As members of an ecosystem it behooves us to consider our neighbors.

File: tests/integration/test.viewadapter.js [hunk 5]
```diff
@@ -0,0 +1,128 @@
+/* To test these adapters, you can run
+ADAPTERS=memory CLIENT=selenium:firefox npm run test */
```
Comment (garbados) [inline tests/integration/test.viewadapter.js:2]: `npm run test` and `npm test` are equivalent.

File: tests/integration/test.viewadapter.js (hunk 3 above)
Comment (ninetteadhikari) [reply tests/integration/test.viewadapter.js:102]: we were trying to see if the view index database is created in Indexeddb (default adapter) and for that we have to run the query.

File: tests/integration/test.viewadapter.js (hunk 1 above)
Comment (ninetteadhikari) [reply tests/integration/test.viewadapter.js:59]: mocha/chai is on version 3.5 here and it uses node 8. This doesn't support ES6 so unfortunately we can't use async/await (thanks @hulkoba on finding this out) 
let me know if there is any other way we can address this issue..

File: tests/integration/test.viewadapter.js (hunk 4 above)
Comment (ninetteadhikari) [reply tests/integration/test.viewadapter.js:58]: good point!I'll combine the two together:)

File: tests/integration/test.viewadapter.js (hunk 5 above)
Comment (ninetteadhikari) [reply tests/integration/test.viewadapter.js:2]: i just added it from here in the document:
https://github.com/neighbourhoodie/pouchdb/blob/master/TESTING.md#L217

**Comment by ninetteadhikari** [comment]
> Just for some context, [other companies](https://github.com/RangerMauve/pouchdb-adapter-hyperbee/issues/1#issuecomment-843623485) have sought this feature and been dissuaded from using PouchDB by its absence. This is part of why I support a broader `dependentDbs` configuration syntax, rather than just a `view_adapter` option. As members of an ecosystem it behooves us to consider our neighbors.

Agreed! as first pass however do you think it would be sufficient to leave it as `view_adapter` and pass a String or Object as needed?  `view_adapter: {"name": "idb", "encryption_key": something"}` and then we can customize it more later.
Maybe @janl or someone else can add more details.

**Comment by garbados** [comment]
> Agreed! as first pass however do you think it would be sufficient to leave it as view_adapter and pass a String or Object as needed? view_adapter: {"name": "idb", "encryption_key": something"} and then we can customize it more later.

I don't believe it makes sense to have the `view_adapter` option be an object, since the `adapter` option is always a string. If we need to pass more options to dependent DBs in order to accomplish our goals, we should go with the `dependentDb` syntax right away. Regardless, I recommend we implement the `dependentDb` syntax prior to submitting this PR to PouchDB itself.

**Comment by janl** [comment]
I don’t think I agree with your comment (sorry to drop a list, I’m about to head off work so I’d like to leave as much context as possible).

1. “dependentDbs” is an implementation concept in PouchDB that is not surfaced to the API anywhere, so I don’t think that’s a better name (n.b. it’d have to be dependent_dbs to match naming conventions
2. view_adapter re-uses two already known concepts from the API. I could be persuaded to bikeshed to this to index_adapter or similar to make clear it also applies to find() indexes (on not-indexeddb adapters).
3. overloading an option to be a string-or-object is a normal thing to do in JS APIs, and I’m not too worried here. Keeps it simple for the simple case and convenient enough for the complex case at the expense of three more lines in the impl.

**Comment by garbados** [comment]
> I could be persuaded

I suppose that settled it then. I'll make it happen.

**Pull request closed without merging**